### PR TITLE
[libc][math] Add scalbln{,f,l,f128} math functions

### PR DESCRIPTION
**Pull request by aaryanshukla**

### CODE REVIEW

File: libc/config/linux/aarch64/entrypoints.txt [hunk 1]
```diff
@@ -718,6 +721,7 @@ if(LIBC_TYPES_HAS_FLOAT128)
     libc.src.math.remquof128
     libc.src.math.rintf128
     libc.src.math.roundevenf128
+    libc.src.math.scalblnf128
     libc.src.math.roundf128
     libc.src.math.scalbnf128
```
Comment (overmighty) [inline libc/config/linux/aarch64/entrypoints.txt:726]: Nit: sorting.

File: libc/newhdrgen/yaml/math.yaml [hunk 2]
```diff
@@ -2199,7 +2206,44 @@ functions:
       - type: float128
       - type: int
       - type: unsigned int
-    guard: LIBC_TYPES_HAS_FLOAT128
+    guard: LIBC_TYPES_HAS_FLOAT128scalbln
```
Comment (overmighty) [inline libc/newhdrgen/yaml/math.yaml:2209]: This line should be reverted.

File: libc/newhdrgen/yaml/math.yaml [hunk 3]
```diff
@@ -2199,7 +2206,44 @@ functions:
...
+    standards:
+      - stdc
+    return_type: float16
+    arguments:
+      - type: float16
+      - type: long
+    guard: LIBC_HAS_TYPES_FLOAT16
+  - name: scalblnf128
+    standards:
+      - llvm_libc_ext
```
Comment (overmighty) [inline libc/newhdrgen/yaml/math.yaml:2241]: All the `scalbln*` functions are standard C functions. They're not LLVM libc extensions.

You can use https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3096.pdf if you need access to a draft of the C23 standard. You can also look up pre-C23 functions on cppreference.com.

File: libc/spec/llvm_libc_ext.td [hunk 4]
```diff
@@ -100,6 +100,9 @@ def LLVMLibcExt : StandardSpec<"llvm_libc_ext"> {
 
           FunctionSpec<"powi", RetValSpec<DoubleType>, [ArgSpec<DoubleType>, ArgSpec<IntType>]>,
           FunctionSpec<"powif", RetValSpec<FloatType>, [ArgSpec<FloatType>, ArgSpec<IntType>]>,
+
+          GuardedFunctionSpec<"scalblnf128", RetValSpec<Float128Type>, [ArgSpec<Float128Type>, ArgSpec<LongType>], "LIBC_TYPES_HAS_FLOAT128">,
+
```
Comment (overmighty) [inline libc/spec/llvm_libc_ext.td:105]: Should be moved to libc/spec/stdc.td.

File: libc/spec/stdc.td [hunk 5]
```diff
@@ -711,7 +711,10 @@ def StdC : StandardSpec<"stdc"> {
           FunctionSpec<"acoshf", RetValSpec<FloatType>, [ArgSpec<FloatType>]>,
           FunctionSpec<"asinhf", RetValSpec<FloatType>, [ArgSpec<FloatType>]>,
           FunctionSpec<"atanhf", RetValSpec<FloatType>, [ArgSpec<FloatType>]>,
-
+
```
Comment (overmighty) [inline libc/spec/stdc.td:714]: Nit: you added trailing spaces in this line.

File: libc/src/math/generic/CMakeLists.txt [hunk 6]
```diff
@@ -4098,6 +4098,47 @@ add_entrypoint_object(
...
+    ../scalbln.h
+  DEPENDS
+    libc.hdr.float_macros
+    libc.src.__support.FPUtil.manipulation_functions
+  COMPILE_OPTIONS
+    -O3
+)
+
+add_entrypoint_object(
+  scalblnl
```
Comment (overmighty) [inline libc/src/math/generic/CMakeLists.txt:4116]: Nit: sorting. `scalblnf` should appear before `scalblnl`.

File: libc/src/math/generic/CMakeLists.txt [hunk 7]
```diff
@@ -4098,6 +4098,47 @@ add_entrypoint_object(
...
+add_entrypoint_object(
+  scalblnf
+  SRCS
+    scalblnf.cpp
+  HDRS
+    ../scalblnf.h
+  DEPENDS
+    libc.hdr.float_macros
+    libc.src.__support.macros.properties.types
+    libc.src.__support.FPUtil.manipulation_functions
```
Comment (overmighty) [inline libc/src/math/generic/CMakeLists.txt:4137]:
```suggestion
  DEPENDS
    libc.hdr.float_macros
    libc.src.__support.FPUtil.manipulation_functions
```

File: libc/src/math/generic/scalblnf.cpp [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/math/scalblnf.h"
+#include "hdr/float_macros.h"
+#include "src/__support/FPUtil/ManipulationFunctions.h"
+#include "src/__support/common.h"
+#include "src/__support/macros/config.h"
+#include "src/__support/macros/properties/types.h"
```
Comment (overmighty) [inline libc/src/math/generic/scalblnf.cpp:14]:
```suggestion
```

File: libc/src/math/generic/scalblnf.cpp [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+#if FLT_RADIX != 2
+#error "FLT_RADIX != 2 is not supported."
+#endif
+namespace LIBC_NAMESPACE_DECL {
```
Comment (overmighty) [inline libc/src/math/generic/scalblnf.cpp:19]: Nit: whitespace.

```suggestion
#endif

namespace LIBC_NAMESPACE_DECL {
```

File: libc/src/math/scalblnl.h [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_SRC_MATH_SCALBLNL_H
+#define LLVM_LIBC_SRC_MATH_SCALBLNL_H
+
+#include "src/__support/macros/config.h"
+#include "src/__support/macros/properties/types.h"
```
Comment (overmighty) [inline libc/src/math/scalblnl.h:13]:
```suggestion
```

File: libc/test/src/math/smoke/CMakeLists.txt [hunk 11]
```diff
@@ -3618,6 +3618,45 @@ add_fp_unittest(
...
+  SUITE
+    libc-math-smoke-tests
+  SRCS
+    scalbln_test.cpp
+  HDRS
+    ScalbnTest.h
+  DEPENDS
+    libc.src.math.scalbln
+    libc.src.__support.FPUtil.fp_bits
+)
```
Comment (overmighty) [inline libc/test/src/math/smoke/CMakeLists.txt:3632]: The `scalbn*` tests' dependencies should be updated to match the includes of LdExpTest.h, which is itself included by ScalbnTest.h. The `ldexp*` tests' dependencies are correct.

File: libc/newhdrgen/yaml/math.yaml [hunk 12]
```diff
@@ -1845,6 +1845,13 @@ functions:
       - type: float128
       - type: float128
     guard: LIBC_TYPES_HAS_FLOAT128
+  - name: floorf128
```
Comment (lntue) [inline libc/newhdrgen/yaml/math.yaml:1848]: should this be in a separate PR?

**Review by overmighty — CHANGES_REQUESTED** [review]
I would drop C23 from the PR title, as the `scalbln` function family is from C99. See https://en.cppreference.com/w/c/numeric/math/scalbn. Only `scalblnf128` is a C23 function, so I don't think it would make sense to have C23 in the PR title.

File: libc/newhdrgen/yaml/math.yaml (hunk 3 above)
Comment (lntue) [inline libc/newhdrgen/yaml/math.yaml:2241]: `scalblnf128` should be in the stdc standards, not an LLVM extension.

File: libc/spec/llvm_libc_ext.td [hunk 13]
```diff
@@ -100,6 +100,9 @@ def LLVMLibcExt : StandardSpec<"llvm_libc_ext"> {
 
           FunctionSpec<"powi", RetValSpec<DoubleType>, [ArgSpec<DoubleType>, ArgSpec<IntType>]>,
           FunctionSpec<"powif", RetValSpec<FloatType>, [ArgSpec<FloatType>, ArgSpec<IntType>]>,
+
+          GuardedFunctionSpec<"scalblnf128", RetValSpec<Float128Type>, [ArgSpec<Float128Type>, ArgSpec<LongType>], "LIBC_TYPES_HAS_FLOAT128">,
```
Comment (lntue) [inline libc/spec/llvm_libc_ext.td:104]: move this to `stdc.td`

File: libc/newhdrgen/yaml/math.yaml [hunk 14]
```diff
@@ -1898,7 +1898,7 @@ functions:
     return_type: float128
     arguments:
       - type: float128
-    guard: LIBC_TYPES_HAS_FLOAT128
+    guard: LIBC_TYPES_HAS_FLOAT12
```
Comment (overmighty) [inline libc/newhdrgen/yaml/math.yaml:1901]: Regression.

File: libc/spec/stdc.td [hunk 15]
```diff
@@ -712,7 +712,11 @@ def StdC : StandardSpec<"stdc"> {
           FunctionSpec<"asinhf", RetValSpec<FloatType>, [ArgSpec<FloatType>]>,
           FunctionSpec<"atanhf", RetValSpec<FloatType>, [ArgSpec<FloatType>]>,
 
+          FunctionSpec<"scalbln", RetValSpec<DoubleType>, [ArgSpec<DoubleType>, ArgSpec<LongType>]>,
+          FunctionSpec<"scalblnl", RetValSpec<LongDoubleType>, [ArgSpec<LongDoubleType>, ArgSpec<LongType>]>,
+          FunctionSpec<"scalblnf", RetValSpec<FloatType>, [ArgSpec<FloatType>, ArgSpec<LongType>]>,
```
Comment (overmighty) [inline libc/spec/stdc.td:717]: Nit: sorting.

```suggestion
          FunctionSpec<"scalblnf", RetValSpec<FloatType>, [ArgSpec<FloatType>, ArgSpec<LongType>]>,
          FunctionSpec<"scalblnl", RetValSpec<LongDoubleType>, [ArgSpec<LongDoubleType>, ArgSpec<LongType>]>,
```

File: libc/src/math/CMakeLists.txt [hunk 16]
```diff
@@ -419,7 +419,11 @@ add_math_entrypoint_object(roundevenl)
 add_math_entrypoint_object(roundevenf16)
 add_math_entrypoint_object(roundevenf128)
 
+add_math_entrypoint_object(scalbln)
+add_math_entrypoint_object(scalblnl)
+add_math_entrypoint_object(scalblnf)
```
Comment (overmighty) [inline libc/src/math/CMakeLists.txt:424]: Nit: sorting.

```suggestion
add_math_entrypoint_object(scalblnf)
add_math_entrypoint_object(scalblnl)
```

File: libc/src/math/generic/CMakeLists.txt (hunk 7 above)
Comment (overmighty) [reply libc/src/math/generic/CMakeLists.txt]: This wasn't resolved, you just swapped `scalblnl` and `scalblnf`.

File: libc/test/src/math/smoke/CMakeLists.txt [hunk 17]
```diff
@@ -1633,6 +1633,7 @@ add_fp_unittest(
     libc.src.__support.CPP.limits
     libc.src.__support.FPUtil.fp_bits
     libc.src.__support.FPUtil.normal_float
+
```
Comment (overmighty) [inline libc/test/src/math/smoke/CMakeLists.txt:1636]: Nit: extra blank line.

File: libc/test/src/math/smoke/CMakeLists.txt [hunk 18]
```diff
@@ -3618,6 +3619,51 @@ add_fp_unittest(
...
+  SRCS
+    scalbln_test.cpp
+  HDRS
+    ScalbnTest.h
+  DEPENDS
+    libc.src.math.scalbln
+    libc.src.__support.FPUtil.fp_bits
+    libc.src.__support.CPP.limits
+    libc.src.__support.FPUtil.normal_float
+)
```
Comment (overmighty) [inline libc/test/src/math/smoke/CMakeLists.txt:3635]: Nit: sorting. Same for the other changes below.

Here's a shell function I use to check sorting sometimes:

```sh
foo() { cat > /tmp/foo.txt; diff -u /tmp/foo.txt <(sort /tmp/foo.txt) }
```

Run `foo` (feel free to rename it), paste the dependencies or entrypoints or whatever, hit Enter, then hit Ctrl+D, and you'll get something like this:

```sh
~ ➜ foo() { cat > /tmp/foo.txt; diff -u /tmp/foo.txt <(sort /tmp/foo.txt) }
~ ➜ foo
    libc.src.math.scalbln
    libc.src.__support.FPUtil.fp_bits
    libc.src.__support.CPP.limits
    libc.src.__support.FPUtil.normal_float
--- /tmp/foo.txt	2024-08-08 00:19:32.174274463 +0200
+++ /proc/self/fd/11	2024-08-08 00:19:35.004294971 +0200
@@ -1,4 +1,4 @@
     libc.src.math.scalbln
-    libc.src.__support.FPUtil.fp_bits
     libc.src.__support.CPP.limits
+    libc.src.__support.FPUtil.fp_bits
     libc.src.__support.FPUtil.normal_float
~ ➜
```

**Review by overmighty — CHANGES_REQUESTED** [review]

File: libc/newhdrgen/yaml/math.yaml [hunk 19]
```diff
@@ -2200,6 +2200,43 @@ functions:
...
+      - type: float
+      - type: long
+  - name: scalblnf16
+    standards:
+      - stdc
+    return_type: float16
+    arguments:
+      - type: float16
+      - type: long
+    guard: LIBC_HAS_TYPES_FLOAT16
```
Comment (overmighty) [inline libc/newhdrgen/yaml/math.yaml:2231]:
```suggestion
    guard: LIBC_TYPES_HAS_FLOAT16
```

File: libc/newhdrgen/yaml/math.yaml [hunk 20]
```diff
@@ -2200,6 +2200,43 @@ functions:
...
+  - name: scalblnf128
+    standards:
+      - stdc
+    return_type: float128
+    arguments:
+      - type: float128
+      - type: long
+    guard: LIBC_HAS_TYPES_FLOAT128
```
Comment (overmighty) [inline libc/newhdrgen/yaml/math.yaml:2239]:
```suggestion
    guard: LIBC_TYPES_HAS_FLOAT128
```

**Review by overmighty — CHANGES_REQUESTED** [review]

File: libc/test/src/math/smoke/CMakeLists.txt [hunk 21]
```diff
@@ -3618,6 +3618,51 @@ add_fp_unittest(
...
+  SRCS
+    scalblnf_test.cpp
+  HDRS
+    ScalbnTest.h
+  DEPENDS
+    libc.src.math.scalblnf
+    libc.src.__support.FPUtil.fp_bits
+    libc.src.__support.CPP.limits
+    libc.src.__support.FPUtil.normal_float
+)
```
Comment (overmighty) [inline libc/test/src/math/smoke/CMakeLists.txt:3649]: Nit: sorting.

```suggestion
  DEPENDS
    libc.src.math.scalblnf
    libc.src.__support.CPP.limits
    libc.src.__support.FPUtil.fp_bits
    libc.src.__support.FPUtil.normal_float
)
```

**Review by overmighty — APPROVED** [review]
LGTM with one more nit.

**Review by lntue — APPROVED** [review]

**Pull request merged**